hot-reload.md

```diff
@@ -43,74 +44,85 @@ app continues to execute from where it was prior to running the hot reload
+Dart code to keep using hot reload.
+
+## Previous state is combined with new code 
+
+Flutter's hot reload feature, sometimes described as _stateful hot reload_,
+preserves the state of your app. This design enables you to view
+the effect of the most recent change only, without throwing away the
+current state. For example, if your app requires a user to log in, you can
+modify and hot reload a page several levels down in the navigation hierarchy,
+without re-entering your login credentials. State is kept, which is 
+usually the desired behavior.
+
+If code changes affect the state of your app (or its dependencies), 
+the data your app has to work with might not be fully consistent with 
+the data it would have if it executed from scratch. The result might be 
+different behavior after hot reload versus a full restart.
 
-Even when a hot reload operation appears successful and generates no exceptions,
-some code changes might not be visible in the refreshed UI. This behavior is
-common after changes to the app's `main()` method.
+For example, if you modify a class definition from extending StatelessWidget
+to StatefulWidget (or the reverse), after hot reload the previous state of
+your app is preserved. However, the state might not be compatible with the
+new changes.
 
-As a general rule, if the modified code is downstream of the root widget's
-build method, then hot reload behaves as expected. However, if the modified code
-won't be re-executed as a result of rebuilding the widget tree, then you won't
-see its effects after hot reload.
+Consider the following code:
 
-For example, consider the following code:
 ```
-import 'package:flutter/material.dart';
-
-void main() {
-  runApp(new MyApp());
-}
-
-class MyApp extends StatelessWidget {
+class myWidget extends StatelessWidget {
   Widget build(BuildContext context) {
-    return new GestureDetector(
-      onTap: () => print('tapped'));
+    return new GestureDetector(onTap: () => print('T'));
   }
 }
 ```
-
-After running this app, you might change the code as follows:
+After running the app, if you make the following change:
 
 ```
-import 'package:flutter/widgets.dart';
-void main() {
-  runApp(
-    const Center(
-      child: const Text('Hello', textDirection: TextDirection.ltr)));
-  }
+class myWidget extends StatefulWidget {
+  @override
+  State createState() => new myWidgetState();
+}
+class myWidgetState {
+...
+...
+}
 ```
 
+and then hot reload, the console displays an assertion failure similar to:
 
-With a full restart, the program starts from the beginning, executes the new
-version of `main()`, and builds a widget tree that displays the text `Hello`.
-
-However, if you hot reload the app after this change, `main()` is not 
-re-executed, and the widget tree is rebuilt with the unchanged instance of 
-`MyApp` as the root widget. The result is no visible change after hot reload.
-
+```
+myWidget is not a subtype of StatelessWidget
+```
 
```

Comment:
I think we need to add a comment under this failure message along the lines of "In these cases, issue a full restart to see the updated application."